lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py

```diff
@@ -0,0 +1,296 @@
+        return self._datacenter
+
+    @property
+    def datastore(self):
+        ''' Returns the Datastore managed object. '''
+        return self._datastore
+
+    @property
+    def debounce(self):
+        ''' Returns true if the upload should be debounced. '''
+        return self.params['debounce'] != 'none'
+
+    @property
+    def debounce_cache_local(self):
+        ''' Returns true if the local file hash should be persisted. '''
+        return self.params['debounce'] == 'hashcache'
+
+    @property
+    def destination_path(self):
+        ''' Returns the destination path parameter '''
+        return self.params['destination']
+
+    @property
+    def destination_sha256_path(self):
+        ''' Returns the destination path to the sha256 file '''
+        return self.destination_path + '.sha256'
+
+    @property
+    def source_path(self):
+        ''' Returns the source path parameter '''
+        return self.params['source']
+
+    @property
+    def source_sha256_path(self):
+        ''' Returns the source path to the sha256 file '''
+        return self.source_path + '.sha256'
+
+    def upload(self):
+        ''' Upload.  If a host is unreachable then try another. '''
+        for host_mount in self.datastore.host:
+            try:
+                self._esxi = host_mount.key
+                if self.debounce:
+                    same, hash = self._destination_exists()
+                    if same:
+                        self.module.exit_json(changed=False)
+                self._put(self.destination_path, self.data)
+                if self.debounce:
+                    self._put(self.destination_sha256_path, hash)
+                self.module.exit_json(changed=True)
+            except Exception as e:
+                self._history.append(to_native(e))
+        self.module.fail_json(msg='Unable to upload.', reason=self._history)
+
+    def _destination_exists(self):
+        '''
+            Check to see if the file already exists and has a corresponding
+            .sha256 file on the destination.  If so, this operation can be
+            idempotent by reading the .sha256 file, and calculating the local
+            SHA256, and if they match then there is nothing to do.
+
+            Returns:
+                ( same, hash )
+        '''
+        local_hash = self._local_hash()
+        self.module.debug('LOCAL  SHA256: ' + local_hash)
+        remote_hash = self._remote_hash()
+        self.module.debug('REMOTE SHA256: ' + str(remote_hash))
+        return (local_hash == remote_hash, local_hash)
+
+    def _expand(self, url, params):
+        ''' Returns the complete url, needed to request a ticket. '''
+        return requests.Request('GET', url, params=params).prepare().url
+
+    def _get(self, url, headers=None, params=None):
+        ''' Issue a GET - on 404 response return is None otherwise text '''
+        r = requests.get(url, headers=headers, params=params, verify=self.module.params['validate_certs'])
```

Comment:
Use https://github.com/ansible/ansible/blob/cd988f645aaf774c55a98a90e3ef42cc5b1a1563/lib/ansible/module_utils/urls.py#L1147 instead of requests.get

Reply:
I avoided `module_utils/urls.py` for a number of reasons - it appears that a convenience function called `fetch_url` exists to make it a little easier to issue a `GET`, but there is no corresponding `PUT` wrapper.  `requests` is used liberally throughout other modules like google's, and it's much easier to deal with.  Is your comment above a requirement for the vmware modules?